Upgrade to `svgo` version 4+

xavierfoucrier opened the pull request:
Hello @cascornelissen 👋 

This is a pull request related to #214, mainly contains:
- dependency updates / upgrades
- esm code refactoring
- support for svgo 4+

> Currently, this is working well with `svgo@3.3.2` but the goal here is to directly target the latest version of `svgo`, I will continue working on this in the next few days/weeks and will ping you when everything is setup for version 4+ 🎉

cascornelissen commented:
Hi @xavierfoucrier, thanks a bunch for all of these contributions. Much appreciated! ❤️ Is there any chance we can split these up into multiple PRs? 26 commits and 60+ files changed is a nightmare to review properly 😅 

Also, don't worry too much about these minor/patch version bumps of all the dependencies. I tend to update every minor/patch release right before creating a release of this plugin.

xavierfoucrier commented:
Hello @cascornelissen 👋 

Thanks, it's a pleasure 😄 
And sorry, I can understand ... 🙄 

There was some **breaking changes** with dependencies, that's why I updated / upgraded most of them, step by step, to check if tests was still passing well, then move forward in your code to adapt it 👌

Even if I split it in multiple PRs, you will still need to review the **61 file changes** due to the commit **[♻️ refactor to use esm convention](https://github.com/cascornelissen/svg-spritemap-webpack-plugin/pull/220/commits/5f5fda481d2e84dc215ecaac561d5c777676df81)**, which refactor all imports/exports to use the ESM modern convention instead of CJS. For me this is just the hardest part, but clearly should be fast because it's always (almost) the same renaming.

For the rest:
- multiple package updates and upgrades without any other code changes
- `import` was adapted in **[⬆️ upgrade svg-element-attributes to 2.1.0](https://github.com/cascornelissen/svg-spritemap-webpack-plugin/pull/220/commits/c27d43c699efd6a310dd0e3ad1c46585d060dd87)**
- output assets was adapted due to `svgo` compression in **[📁 update output svg assets based on svgo compression](https://github.com/cascornelissen/svg-spritemap-webpack-plugin/pull/220/commits/1a0dce2dc0fb236a9b38e01468b888298a044f7b)**
- `cleanupIDs` has been renamed by `svgo` to `cleanupIds` in **[✏️ rename cleanupIds plugin](https://github.com/cascornelissen/svg-spritemap-webpack-plugin/pull/220/commits/5275ab0ae7199c8f8d750122718288fb1e0071b2)**, see https://svgo.dev/docs/plugins/cleanupIds/

And right now, I haven't finished implementing totally the v4, because of some issues regarding `svgo` configuration.

I can split the work in **two parts**, as for the moment the plugin is working fine at least with `svgo@3.3.2`: may be you would like to push a new version for the v3 while waiting for the final release of v4?

Let me know 😊

cascornelissen commented:
Thanks for the thorough explanation! As mentioned I'm not planning on doing a separate release for SVGO 3 and 4 as they would both likely result in a major release for this plugin as well.

I should've probably asked it more explicitly: is there a need to switch to ESM to update to SVGO 3 (or 4, later on)? As mentioned, I'm fine with both changes but doing everything in one big bang is more likely to lead to issues and in the end it's my responsibility 😅 It's not just more work on my side for the reviewing, it will likely also lead to more back-and-forth as you'll receive more comments on different parts of the PR.

xavierfoucrier commented:
From what I have read in releases, `svgo` v3 was refactored to ESM, but they finally keep support of CJS, see https://github.com/svg/svgo/releases.

Since you have planned to release a major version of the plugin, I think it should be good to upgrade to ESM for many reasons:
- Align with JavaScript standards
- Native support in browsers
- Better tree-shaking and optimization
- Improved asynchronous loading
- Enhanced compatibility
- Consistency across ecosystem
- Future proofing
- Better Interoperability with TypeScript
- and so on...

But, it you think it's better to let it as is right now, I can rollback this part and see if everything goes fine with `svgo`.

cascornelissen commented:
Not sure where the miscommunication lies but I'm trying to say that I am interested in both updating to SVGO3/4 _and_ switching to ESM. But refactoring an entire project in a single PR is a recipe for disaster. That's why I'm wondering if it's possible to create a few separate PRs;

1. Switch to ESM
2. Add support for SVGO 3/4

This way we could even do a minor release for the ESM implementation, assuming it's backwards compatible.

xavierfoucrier commented:
@cascornelissen 👋 

Ha sorry, probably my side, tired these past few days 😄
I will split the work in two parts, no problem!

Agree with you 😉
